File: backend/api/routers/backup.py
from __future__ import annotations

from fastapi import APIRouter, Depends

from backend.api.dependencies import get_container
from backend.api.schemas.backup import (
    CreateBackupPlanRequest,
    EvaluateRetentionRequest,
    ErrorPayload,
    ResponseEnvelope,
    RestoreBackupRequest,
    RunBackupRequest,
    UpdateBackupPlanRequest,
)
from backend.application.backup import BackupApplicationError
from backend.domain.shared_kernel import ProjectId
from backend.infrastructure.di import ApplicationContainer


router = APIRouter(prefix="/api/v1", tags=["backup"])


@router.get("/projects/{project_id}/backups/plans", response_model=ResponseEnvelope)
def list_backup_plans(project_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    return _success(container.create_backup_service().list_plans(ProjectId(project_id)))


@router.post("/projects/{project_id}/backups/plans", response_model=ResponseEnvelope)
def create_backup_plan(
    project_id: str,
    request: CreateBackupPlanRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_backup_service().create_plan(
                ProjectId(project_id),
                name=request.name,
                backup_scope=request.backup_scope,
                retention_policy=request.retention_policy,
                schedule_interval_seconds=request.schedule_interval_seconds,
                is_enabled=request.is_enabled,
            )
        )
    except BackupApplicationError as error:
        return _failure(error)


@router.patch("/projects/{project_id}/backups/plans/{plan_id}", response_model=ResponseEnvelope)
def update_backup_plan(
    project_id: str,
    plan_id: str,
    request: UpdateBackupPlanRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_backup_service().update_plan(
                ProjectId(project_id),
                plan_id=plan_id,
                retention_policy=request.retention_policy,
                schedule_interval_seconds=request.schedule_interval_seconds,
                is_enabled=request.is_enabled,
            )
        )
    except BackupApplicationError as error:
        return _failure(error)
@router.get("/projects/{project_id}/backups/runs", response_model=ResponseEnvelope)
def list_backup_runs(project_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    return _success(container.create_backup_service().list_runs(ProjectId(project_id)))


@router.get("/projects/{project_id}/backups/runs/{backup_run_id}", response_model=ResponseEnvelope)
def get_backup_run(
    project_id: str,
    backup_run_id: str,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(container.create_backup_service().get_run(ProjectId(project_id), backup_run_id))
    except BackupApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/backups/runs", response_model=ResponseEnvelope)
def run_backup(
    project_id: str,
    request: RunBackupRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        result = container.create_backup_service().execute_run_backup(
            ProjectId(project_id),
            plan_id=request.backup_plan_id,
            idempotency_key=request.idempotency_key,
        )
        return _success(result, warnings=result.get("warnings", []))
    except BackupApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/backups/restores/plan", response_model=ResponseEnvelope)
def plan_restore(
    project_id: str,
    request: RestoreBackupRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        preview = container.create_backup_service().preview_restore(ProjectId(project_id), request.backup_run_id)
        return _success(preview, warnings=preview.get("warnings", []))
    except BackupApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/backups/restores", response_model=ResponseEnvelope)
def execute_restore(
    project_id: str,
    request: RestoreBackupRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_backup_service().execute_restore(
                ProjectId(project_id),
                request.backup_run_id,
                confirm_destructive=request.confirm_destructive,
                idempotency_key=request.idempotency_key,
            )
        )
    except BackupApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/backups/restores/{restore_run_id}/undo", response_model=ResponseEnvelope)
def undo_restore(
    project_id: str,
    restore_run_id: str,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(container.create_backup_service().undo_restore(ProjectId(project_id), restore_run_id))
    except BackupApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/backups/retention/evaluate", response_model=ResponseEnvelope)
def evaluate_retention(
    project_id: str,
    request: EvaluateRetentionRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_backup_service().evaluate_retention(
                ProjectId(project_id),
                plan_id=request.backup_plan_id,
            )
        )
    except BackupApplicationError as error:
        return _failure(error)


def _success(data: dict | list[dict], *, warnings: list[str] | None = None) -> ResponseEnvelope:
    return ResponseEnvelope(ok=True, data=data, warnings=warnings or [])


def _failure(error: BackupApplicationError) -> ResponseEnvelope:
    return ResponseEnvelope(
        ok=False,
        error=ErrorPayload(code=error.code.value, message=str(error)),
    )

File: frontend/src/api/backup.ts
import { jsonRequest, requestBackend, type BackendResponse } from "./backend";

export interface BackupPlan {
  backup_plan_id: string;
  project_id: string;
  name: string;
  backup_scope: string;
  retention_policy: { keep_count?: number; keep_days?: number } | null;
  schedule_interval_seconds: number | null;
  next_run_at: string | null;
  is_enabled: boolean;
}

export interface BackupRun {
  backup_run_id: string;
  backup_plan_id: string | null;
  project_id: string;
  status: string;
  trigger_type: string;
  started_at: string;
  finished_at: string | null;
  total_bytes: number | null;
  content_hash: string | null;
  archive_path: string | null;
  manifest_json: Record<string, unknown> | null;
  idempotency_key: string | null;
}

export interface RestorePreview {
  backup_run_id: string;
  project_root: string;
  overwrite_paths: string[];
  warnings: string[];
  requires_pre_restore_snapshot: boolean;
}

export interface RestoreResult {
  restore_run_id: string;
  status: string;
  command_execution_id?: string;
  pre_restore_snapshot_path?: string | null;
  undo_available?: boolean;
}

export interface CreateBackupPlanRequest {
  name: string;
  backup_scope?: string;
  retention_policy?: { keep_count?: number; keep_days?: number } | null;
  schedule_interval_seconds?: number | null;
  is_enabled?: boolean;
}

export type BackupResponse<T> = BackendResponse<T>;

export async function listBackupPlans(projectId: string): Promise<BackupPlan[]> {
  return (await requestBackend<BackupPlan[]>(`/api/v1/projects/${projectId}/backups/plans`)).data;
}

export async function createBackupPlan(projectId: string, request: CreateBackupPlanRequest): Promise<BackupPlan> {
  return (await requestBackend<BackupPlan>(`/api/v1/projects/${projectId}/backups/plans`, jsonRequest(request))).data;
}

export interface UpdateBackupPlanRequest {
  retention_policy?: { keep_count?: number; keep_days?: number } | null;
  schedule_interval_seconds?: number | null;
  is_enabled?: boolean | null;
}

export async function updateBackupPlan(projectId: string, planId: string, request: UpdateBackupPlanRequest): Promise<BackupPlan> {
  return (await requestBackend<BackupPlan>(`/api/v1/projects/${projectId}/backups/plans/${planId}`, {
    method: "PATCH",
    ...jsonRequest(request)
  })).data;
}

export async function listBackupRuns(projectId: string): Promise<BackupRun[]> {
  return (await requestBackend<BackupRun[]>(`/api/v1/projects/${projectId}/backups/runs`)).data;
}

export async function getBackupRun(projectId: string, backupRunId: string): Promise<BackupRun & { items?: Record<string, unknown>[] }> {
  return (await requestBackend<BackupRun & { items?: Record<string, unknown>[] }>(`/api/v1/projects/${projectId}/backups/runs/${backupRunId}`)).data;
}

export async function runBackup(projectId: string, backupPlanId?: string | null): Promise<BackendResponse<BackupRun>> {
  return requestBackend<BackupRun>(
    `/api/v1/projects/${projectId}/backups/runs`,
    jsonRequest({ backup_plan_id: backupPlanId ?? null, idempotency_key: crypto.randomUUID() })
  );
}

export async function previewRestore(projectId: string, backupRunId: string): Promise<BackupResponse<RestorePreview>> {
  const response = await requestBackend<RestorePreview>(`/api/v1/projects/${projectId}/backups/restores/plan`, jsonRequest({ backup_run_id: backupRunId }));
  return response;
}

export async function executeRestore(
  projectId: string,
  backupRunId: string,
  confirmDestructive: boolean
): Promise<RestoreResult> {
  return (
    await requestBackend<RestoreResult>(
      `/api/v1/projects/${projectId}/backups/restores`,
      jsonRequest({ backup_run_id: backupRunId, confirm_destructive: confirmDestructive, idempotency_key: crypto.randomUUID() })
    )
  ).data;
}

export async function undoRestore(projectId: string, restoreRunId: string): Promise<Record<string, unknown>> {
  return (await requestBackend<Record<string, unknown>>(`/api/v1/projects/${projectId}/backups/restores/${restoreRunId}/undo`, jsonRequest({}))).data;
}

export async function evaluateRetention(projectId: string, backupPlanId?: string | null): Promise<{ pruned: string[]; skipped: string[]; evaluated: number }> {
  return (
    await requestBackend<{ pruned: string[]; skipped: string[]; evaluated: number }>(
      `/api/v1/projects/${projectId}/backups/retention/evaluate`,
      jsonRequest({ backup_plan_id: backupPlanId ?? null })
    )
  ).data;
}
